lib/spack/docs/tutorial_configuration.rst

```diff
@@ -0,0 +1,700 @@
+    mpich:
+      paths:
+        mpich@3.2%gcc@5.4.0 device=ch3 +hydra netmod=tcp +pmi+romio~verbs arch=linux-ubuntu16.04-x86_64: /usr
+      buildable: False
+
+If we concretize `hdf5+mpi` with this configuration file, we will just
+build with an alternate MPI implementation.
+
+.. code-block:: console
+
+  $ spack spec hdf5 %clang +mpi
+  Input spec
+  --------------------------------
+  hdf5%clang+mpi
+
+  Normalized
+  --------------------------------
+  hdf5%clang+mpi
+      ^mpi
+      ^zlib@1.1.2:
+
+  Concretized
+  --------------------------------
+  hdf5@1.10.1%clang@3.8.0-2ubuntu4~cxx~debug~fortran~hl+mpi+pic~shared~szip~threadsafe arch=linux-ubuntu16.04-x86_64
+      ^openmpi@3.0.0%clang@3.8.0-2ubuntu4~cuda fabrics=verbs ~java schedulers= ~sqlite3~thread_multiple+vt arch=linux-ubuntu16.04-x86_64
+          ^hwloc@1.11.8%clang@3.8.0-2ubuntu4~cuda+libxml2+pci arch=linux-ubuntu16.04-x86_64
+              ^libpciaccess@0.13.5%clang@3.8.0-2ubuntu4 arch=linux-ubuntu16.04-x86_64
+                  ^libtool@2.4.6%clang@3.8.0-2ubuntu4 arch=linux-ubuntu16.04-x86_64
+                      ^m4@1.4.18%clang@3.8.0-2ubuntu4 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00 +sigsegv arch=linux-ubuntu16.04-x86_64
+                          ^libsigsegv@2.11%clang@3.8.0-2ubuntu4 arch=linux-ubuntu16.04-x86_64
+                  ^pkg-config@0.29.2%clang@3.8.0-2ubuntu4+internal_glib arch=linux-ubuntu16.04-x86_64
+                  ^util-macros@1.19.1%clang@3.8.0-2ubuntu4 arch=linux-ubuntu16.04-x86_64
+              ^libxml2@2.9.4%clang@3.8.0-2ubuntu4~python arch=linux-ubuntu16.04-x86_64
+                  ^xz@5.2.3%clang@3.8.0-2ubuntu4 arch=linux-ubuntu16.04-x86_64
+                  ^zlib@1.2.11%gcc@5.4.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64
+
+We have only expressed a preference for mpich over other MPI
+implementations, and Spack will happily build with one we haven't
+forbid it from using. We could resolve this by requesting
+`hdf5%clang+mpi^mpich` explicitly, or we can configure Spack not to
+use any other MPI implementation. Since we're focused on
+configurations here and the former can get tedious, we'll need to
+modify our `packages.yaml` file again.
+
+While we're at it, we can configure hdf5 to build with MPI by default
+again.
+
+.. code-block:: yaml
+
+  packages:
+    all:
+      compiler: [clang, gcc, intel, pgi, xl, nag]
+      providers:
+        mpi: [mpich, openmpi]
+      variants: ~shared
+    zlib:
+      paths:
+        zlib@1.2.11%gcc@5.4.0+pic+shared arch=linux-ubuntu16.04-x86_64: /usr
+      buildable: False
+    mpich:
+      paths:
+        mpich@3.2%gcc@5.4.0 device=ch3 +hydra netmod=tcp +pmi+romio~verbs arch=linux-ubuntu16.04-x86_64: /usr
+      buildable: False
+    openmpi:
+      buildable: False
+    mvapich2:
+      buildable: False
+    intel-mpi:
+      buildable: False
+    spectrum-mpi:
+      buildable: False
+    intel-parallel-studio:
+      buildable: False
```

Comment:
Is this seriously the recommended way of working around this bug? If `mpich` is my default MPI, Spack should use `mpich`. If it doesn't, that's a bug.

Reply:
Preferences in the packages.yaml file are always taken as a preference rather than as a constraint. I agree that this should change, but I don't think we're all that close to having the answer.